src/syscall_sdl.c

```diff
@@ -193,34 +208,48 @@ void syscall_draw_frame(struct riscv_t *rv)
     SDL_RenderPresent(renderer);
 }
 
-void syscall_poll_event(struct riscv_t *rv)
+void syscall_setup_queue(struct riscv_t *rv)
 {
-    state_t *s = rv_userdata(rv); /* access userdata */
-
-    /* poll_event(event) */
-    const uint32_t base = rv_get_reg(rv, rv_reg_a0);
-
-    event_t event;
-    if (!event_pop(&event)) {
+    /* setup_queue(capacity, event_count) */
+    queues_capacity = rv_get_reg(rv, rv_reg_a0);
+    event_count = rv_get_reg(rv, rv_reg_a1);
+    if (queues_capacity == 0) {
         rv_set_reg(rv, rv_reg_a0, 0);
         return;
     }
 
-    memory_write(s->mem, base + 0, (const uint8_t *) &event.type, 4);
-    switch (event.type) {
-    case KEY_EVENT:
-        memory_write(s->mem, base + 4, (const uint8_t *) &event.key_event,
-                     sizeof(key_event_t));
-        break;
-    case MOUSE_MOTION_EVENT:
-        memory_write(s->mem, base + 4, (const uint8_t *) &event.mouse.motion,
-                     sizeof(mouse_motion_t));
-        break;
-    case MOUSE_BUTTON_EVENT:
-        memory_write(s->mem, base + 4, (const uint8_t *) &event.mouse.button,
-                     sizeof(mouse_button_t));
-        break;
-    }
-
-    rv_set_reg(rv, rv_reg_a0, 1);
+    /* Bit Twiddling Hacks: Round up to the next highest power of 2 */
```

Comment:
Move the below implementation to a dedicated helper function. In addition, you can use `__builtin_clzl` for the fastpath (clang/gcc) and bitwise hacks as fallback. See https://jameshfisher.com/2018/03/30/round-up-power-2/